Adding support for offscreen render and keyboard/mouse event sending to browser-window.

brenca opened the pull request:
Offscreen render, mouse and keyboard event sending functionality using the content api. This relies on my other pull-request successfully being merged into native-mate, because the offscreen-render callback sends the image data in a Uint8ClampedArray, and I couldn't return that without the things I added to native-mate.

zcbenz commented on atom/browser/native_window.cc:
```diff
@@ -540,6 +573,73 @@ void NativeWindow::DevToolsClosed() {
+  const auto host = view ? view->GetRenderWidgetHost() : nullptr;
+  host->ForwardKeyboardEvent(*keyb_event);
+
+  if(keyb_event->type == blink::WebInputEvent::Type::KeyDown){
```
Why forwarding the key down event for twice?

brenca replied:
Fixed this in an other commit, I was trying to figure out how I can "forge" valid events that produce the characters I wanted, and for some reason I didn't remove the duplicate sending - I checked, it works with a simple `keydown` event.

zcbenz commented:
For the name of the `setOffscreenRender` API, it implies the rendering would not be drew on the window, but in fact it only seems to capture the frame? We should probably rename it to `BeginFrameSubscription` and `EndFrameSubscription`, making the name match the real behavior.

zcbenz commented on atom/browser/api/atom_api_window.cc:
```diff
@@ -92,6 +95,16 @@ void Window::WillCloseWindow(bool* prevent_default) {
+  v8::Locker locker(isolate());
+  v8::HandleScope handle_scope(isolate());
+
+  v8::Local<v8::ArrayBuffer> data = v8::ArrayBuffer::New(isolate(), rgb.get(), size);
```
`blink` requires creating `v8::ArrayBuffer` from its own API, otherwise it would crash when you print it in the console of devtools. We can just use `node::Buffer::New` here, which is safe and more easy to use.

brenca replied:
I didn't know that, thanks! It should be fixed with the next commit.

brenca commented:
I also renamed the js side and the functions associated with it from `setOffscreenRender` to `Begin/EndFrameSubscription`. I also added a property to the `browser-window` before, but I didn't rename that one yet - I haven't come up with a better name yet and I think as a property, `offscreen-render` is less weird than `setOffscreenRender` was, but it's still a bit weird.

brenca commented:
Since now `node::Buffer` is used instead of `v8::ArrayBuffer`, my pull request to native_mate is not needed,

zcbenz commented:
Thanks!

zcbenz commented:
@brenca It seems that the data passed in `frame-rendered` is raw pixel data? How do you usually deal with it?

pergy commented:
afaik we simply fed to texture (in our case Ogre::Texture), which has the same alignment (BGRA) and size.

zcbenz commented:
@pergy This implementation converts the frame to 32bit ARGB, I'm not sure if it is compatible with BGRA. Does it make more sense to pass a [NativeImage](https://github.com/atom/electron/blob/master/docs/api/native-image.md) in the event?

pergy commented:
@zcbenz Of course we can use other alignments on the rendering side to match the raw data features (seems like we just had luck with bgra :P ). NativeImage may be a clearer choice, but I'm afraid, that conversions on any side can really slow down the rendering, although we didn't make detailed tests on that yet.

brenca commented:
We could create an option for this - you either get the image in a `NativeImage` or you get the raw buffer data. We used the raw data because this is faster than converting to an image and using that/converting back on our end to raw data.

I tested the buffer data that we get from the conversion and it seems to be BGRA, even 'tho the header file states it is ARGB.

zcbenz commented:
I think it is fine to pass raw pixel data for now, we can add options in future if there are people asking for that.
